### ISSUE TITLE
When farbling canvas, farble across all channels

### ISSUE THREAD

**Issue by Thorin-Oakenpants** [brave/brave-browser#12069]
**Updated Issue (@pes10k edit)**
Currently the canvas farbling only effects one channel.  That channel is determined by the per-eTLD+1, per-session farbling seed, so the channel isn't guessable, but it is probably detectable given a determined attacker.

It might make reversing the canvas farbling slightly more difficult if the farbling was done across all channels, which would make it a bit more difficult for an attacker.

**Original Issue**
There is a flaw in canvas randomizing. It's not a security issue. Do you want to discuss it here, or via email, or in a private repo. I'm not talking about the per-origin, per-execution to protect the random seed. It should be a simple fix

**Comment by srirambv** [comment]
cc: @pes10k

**Comment by pes10k** [comment]
Howdy @Thorin-Oakenpants , nice to run into you again :)

If you're in anyway not 100% sure that its not a security issue, feel free to email me at pes@brave.com and we can discuss more there.  Otherwise, feel free to add your findings / concerns here and we can sort things out here.

Thanks!

**Comment by Thorin-Oakenpants** [comment]
@pes10k email sent: feel free to close this (or not)

**Comment by pes10k** [comment]
Just to keep things transparent, the concern is about how reversable canvas fingerprinting protections are (answer, not very hard to reverse!). Specifically, whether 

1. it'd be better to not just farble a single channel (determined by the seed), and instead to farble all channels
2. whether it'd be better to not use the current per-session, per-etld approach, but instead do something shorter, like frame length seeds, to prevent tracking 1p tracking w/in the same-session

I'm going to update this issue accordingly:

1. This would be nice to do one, but its pretty low priority
2. I dont think we'll change this for reasons discussed over email (mainly, it makes it easier for the site to get the seed, and undo all farbling protections, not just for canvas)

Thanks again to @Thorin-Oakenpants for reporting!

**Comment by Thorin-Oakenpants** [comment]
Just to be clear: item 2 is out of scope, and was asked as an off-topic generic overall design question: @pes10k 's information in the email helps explain some of those decisions - thanks! I still think there can be changes there, but not in this ticket

**Comment by pes10k** [comment]
Sorry, apologies @Thorin-Oakenpants , didn't mean to put words in your mouth or misstate your comments.  Thank you for clarifying!

**Comment by Thorin-Oakenpants** [comment]
^^ no worries: I see OP has been edited: "but it is probably detectable given a determined attacker" <- just to make that crystal clear - it is 100% detectable (with the right script), not probably

**Comment by Thorin-Oakenpants** [comment]
randomizing can also be completely bypassed in iframes and currently offscreen worker doesn't seem to be covered: @pes10k email me if you want details, or you may already be aware of these

**Comment by pes10k** [comment]
@Thorin-Oakenpants hmm, the iframes example is very surprising, we have tests, but we may have missed something!  Same with workers!  So, if there are further edge cases we need to cover, that would be very welcome!

You can see our manual test cases here too, fwiw (there are of course automated tests we have in CI too), though you'll need to be in nightly to see everything we expect to be covered.  There are somethings tested in these tests that are not yet implemented, but most of it is.

https://dev-pages.brave.software/farbling.html

TL;DR; Yes, more details / sunlight you can offer, the more thats appreciated!

**Comment by Thorin-Oakenpants** [comment]
The methods should affect all your anti-FPing, not just canvas. I'll send you an email

**Comment by pes10k** [comment]
Closing this out because we now also [determine which channel to farble depending on on the seed,](https://github.com/brave/brave-core/blob/ab1d35a83aefd11b2b11eb9ceeab6febf6619ff8/chromium_src/third_party/blink/renderer/core/execution_context/execution_context.cc#L198) which will prevent the attacker from knowing which channels are unmodified and which are farbled.  Thanks for the issue @Thorin-Oakenpants !

**Comment by pes10k** [comment]
@Thorin-Oakenpants pointed out some techniques sites could use to determine which channel is being farbled in the current implementation (thanks!).  After discussing, I'm convinced we should do as the issue suggested, and farble across all channels.  I'll follow up with a possible PR.  Thanks!

**Comment by pes10k** [comment]
for QA, there are new, per channel tests on https://dev-pages.brave.software/farbling.html

**Comment by stephendonner** [comment]
Verified `FIXED` on the `beta` channel using 

Brave | 1.22.51 Chromium: 89.0.4389.72 (Official Build) beta (x86_64)
-- | --
Revision | 3f345f156bfd157bd1bea06310e55f3fb2490359-refs/branch-heads/4389@{#1393}
OS | macOS Version 11.2.2 (Build 20D80)


Verified that the values per channel (∟ red channel, etc.) do all of the following:

1. differ per origin, so `https://dev-pages.brave.software/farbling.html` and `https://dev-pages.bravesoftware.com/farbling.html` have different values when shields are up (this is the “there site” link at the top of the page, for reference)

**Shields up**, `dev-pages.brave.software/farbling.html` vs. `dev-pages.bravesoftware.com/farbling.html`:

<img width="1680" alt="Screen Shot 2021-03-03 at 11 30 31 AM" src="https://user-images.githubusercontent.com/387249/109864764-1b0f9f00-7c18-11eb-9b63-196abbcbdef2.png">

**Shields down**, `dev-pages.brave.software/farbling.html` vs. `dev-pages.bravesoftware.com/farbling.html`:

<img width="1680" alt="Screen Shot 2021-03-03 at 11 33 21 AM" src="https://user-images.githubusercontent.com/387249/109865029-72157400-7c18-11eb-9c8f-3a1316fd1707.png">

2. differ per browser session (so if you quit the browser and come back, you should get different values, on both origins)

**Shields up**, _restart_, `dev-pages.brave.software/farbling.html` vs. `dev-pages.bravesoftware.com/farbling.html`:

<img width="1677" alt="Screen Shot 2021-03-03 at 11 45 48 AM" src="https://user-images.githubusercontent.com/387249/109865386-d801fb80-7c18-11eb-97be-334785be97ed.png">

3. differ from the shields down version (which should be the same cross origin, cross session)

**Shields down**, _restart_, `dev-pages.brave.software/farbling.html` vs. `dev-pages.bravesoftware.com/farbling.html`:

<img width="1676" alt="Screen Shot 2021-03-03 at 11 48 35 AM" src="https://user-images.githubusercontent.com/387249/109865587-11d30200-7c19-11eb-94e6-c185ba120bdd.png">

In addition to the screenshots, I've captured the values, below; SU = Shields up, SD = Shields down

|       | Shields up - brave.software | Shields up - `bravesoftware.com` | Shields down - `brave.software` | Shields down - `bravesoftware.com` | SU, restart, `brave.software` | SU, restart, `bravesoftware.com` | SD, restart, `brave.software` | SD, restart, `bravesoftware.com` |
| ----- | --------------------------- | ------------------------------ | ----------------------------- | -------------------------------- | --------------------------- | ------------------------------ | --------------------------- | ------------------------------ |
| red   | 72009bc3                    | d971e672                       | 68ede59d                      | 68ed59d                          | 5e30861d                    | 4daf9c46                       | 68ed59d                     | 68ed59d                        |
| green | 44a7bda2                    | e633f8ab                       | f8d9ea89                      | f8d7ea89                         | bbc842c7                    | fb0f2ad7                       | f8d7ea89                    | f8d7ea89                       |
| blue  | 391c891f                    | 691556c2                       | a68c1dac                      | a68c1dac                         | 9735d809                    | 4b8da15b                       | a68c1dac                    | a68c1dac                       |

**Comment by srirambv** [comment]
Verification passed on OnePlus 6T with Android 10 running 1.22.51 x64 build
- Verified steps from https://github.com/brave/brave-browser/issues/12069#issuecomment-773542308

<details>
<summary><b>Case 1: Shields Up, same value</b></summary>

![image](https://user-images.githubusercontent.com/17010094/109984950-2a681880-7cfc-11eb-9232-9ecdca5121c4.png) | ![image](https://user-images.githubusercontent.com/17010094/109984966-2f2ccc80-7cfc-11eb-989c-2972de3b541c.png)
-- | --

</details>
<details>
<summary><b>Case 2: Shields Up, different value on restart</b></summary>

![image](https://user-images.githubusercontent.com/17010094/109985034-410e6f80-7cfc-11eb-8e05-570ee1b70928.png) | ![image](https://user-images.githubusercontent.com/17010094/109985482-b2e6b900-7cfc-11eb-87cb-b1b9fa0d2423.png)
-- | --

</details>
<details>
<summary><b>Case 3: Shields Down, same value</b></summary>

![image](https://user-images.githubusercontent.com/17010094/109985314-8894fb80-7cfc-11eb-91a3-1ac796d8ae27.png) | ![image](https://user-images.githubusercontent.com/17010094/109985297-85017480-7cfc-11eb-9994-02a7e48343b1.png)
-- | --

</details>
<details>
<summary><b>Case 2: Shields Down, same value on restart</b></summary>

![image](https://user-images.githubusercontent.com/17010094/109985372-99de0800-7cfc-11eb-8da2-d6c09c6ba8cd.png) | ![image](https://user-images.githubusercontent.com/17010094/109985391-9d718f00-7cfc-11eb-8d9c-852c491a391b.png)
-- | --

</details>
